src/Confluent.Kafka/Producer.cs

```diff
@@ -88,38 +88,15 @@ private void LogCallback(IntPtr rk, int level, string fac, string buf)
             OnLog.Invoke(this, new LogMessage(name, level, fac, buf));
         }
 
+        private readonly Func<string, SafeTopicHandle> topicHandlerFactory;
+
         /// <remarks>
         ///     getKafkaTopicHandle() is now only required by GetMetadata() which still requires that 
         ///     topic is specified via a handle rather than a string name (note that getKafkaTopicHandle() 
         ///     was also formerly required by the ProduceAsync methods). Eventually we would like to 
         ///     depreciate this method as well as the SafeTopicHandle class.
         /// </remarks>
-        private SafeTopicHandle getKafkaTopicHandle(string topic)
-        {
-            if (topicHandles.ContainsKey(topic))
-            {
-                return topicHandles[topic];
-            }
-
-            var topicConfigHandle = SafeTopicConfigHandle.Create();
-            if (topicConfig != null)
-            {
-                topicConfig
-                    .ToList()
-                    .ForEach((kvp) => { topicConfigHandle.Set(kvp.Key, kvp.Value.ToString()); });
-            }
-            topicConfigHandle.Set("produce.offset.report", "true");
-            IntPtr configPtr = topicConfigHandle.DangerousGetHandle();
-
-            // note: there is a possible (benign) race condition here - topicHandle could have already
-            // been created for the topic (and possibly added to topicHandles). If the topicHandle has
-            // already been created, rdkafka will return it and not create another. the call to rdkafka
-            // is threadsafe.
-            var topicHandle = kafkaHandle.Topic(topic, configPtr);
-            topicHandles.Add(topic, topicHandle);
-
-            return topicHandle;
-        }
+        private SafeTopicHandle getKafkaTopicHandle(string topic) => topicHandles.GetOrAdd(topic, topicHandlerFactory);
```

Comment:
I don't think there is any benefit to using this variant of `GetOrAdd`? If that's right, I think using the other variant is better as it reduces the levels of indirection making the code more straightforward.

other than this, this PR looks good to me, and I'll merge after I've verified the integration tests pass (which they currently don't due to a problem unrelated to you).

Reply:
Currently `Producer.topicConfig` gets captured in a lambda, which forces the compiler to emit a new delegate allocation at every `getKafkaTopicHandle` call.
That's why we have to cache the callback manually.